File: 4.8-bc/py/stage-1-flat/app/service/subscription.py
from __future__ import annotations

from datetime import UTC, datetime, timedelta
from uuid import UUID, uuid4

from app.model.subscription import Subscription
from app.repository.subscription import SubscriptionRepo


class SubscriptionService:
    def __init__(self, repo: SubscriptionRepo) -> None:
        self._repo = repo

    async def subscribe(self, user_id: UUID, plan: str) -> Subscription:
        now = datetime.now(UTC)
        sub = Subscription(
            id=uuid4(),
            user_id=user_id,
            plan=plan,
            next_charge_at=now + timedelta(days=30),
            created_at=now,
        )
        await self._repo.create(sub)
        return sub

File: 4.8-bc/go/stage-1-flat/app/service/subscription.go
package service

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-1-flat/app/model"
)

type SubscriptionRepo interface {
	Create(ctx context.Context, s model.Subscription) error
}

type SubscriptionService struct {
	repo SubscriptionRepo
}

func NewSubscriptionService(r SubscriptionRepo) *SubscriptionService {
	return &SubscriptionService{repo: r}
}

func (s *SubscriptionService) Subscribe(ctx context.Context, userID uuid.UUID, plan string) (*model.Subscription, error) {
	now := time.Now().UTC()
	sub := model.Subscription{
		ID:           uuid.New(),
		UserID:       userID,
		Plan:         plan,
		NextChargeAt: now.AddDate(0, 1, 0),
		CreatedAt:    now,
	}
	if err := s.repo.Create(ctx, sub); err != nil {
		return nil, err
	}
	return &sub, nil
}
